swap/src/bitcoin/wallet.rs

```diff
@@ -215,70 +211,110 @@ impl Wallet {
         self.client.lock().await.status_of_script(tx)
     }
 
-    pub async fn watch_until_status<T>(
-        &self,
-        tx: &T,
-        mut status_fn: impl FnMut(ScriptStatus) -> bool,
-    ) -> Result<()>
-    where
-        T: Watchable,
-    {
+    pub fn subscribe_to(&self, tx: impl Watchable + Send + 'static) -> Subscription {
+        let (sender, receiver) = watch::channel(ScriptStatus::Unseen);
+        let client = self.client.clone();
         let txid = tx.id();
 
-        let mut last_status = None;
+        tokio::spawn(async move {
+            let mut last_status = None;
 
-        loop {
-            let new_status = self.client.lock().await.status_of_script(tx)?;
+            loop {
+                tokio::time::sleep(Duration::from_secs(5)).await;
 
-            if Some(new_status) != last_status {
-                tracing::debug!(%txid, "Transaction is {}", new_status);
-            }
-            last_status = Some(new_status);
+                let new_status = match client.lock().await.status_of_script(&tx) {
+                    Ok(new_status) => new_status,
+                    Err(e) => {
+                        tracing::warn!(%txid, "Failed to get status of script: {:#}", e);
+                        return;
+                    }
+                };
 
-            if status_fn(new_status) {
-                break;
+                if Some(new_status) != last_status {
+                    tracing::debug!(%txid, "Transaction is {}", new_status);
+                }
+                last_status = Some(new_status);
+
+                let all_receivers_gone = sender.send(new_status).is_err();
```

Comment:
❓ Previously this loop was ended upon a certain status being reached. Now we exit by returning if there are no subscribers (i.e. all receivers were dropped) anymore. Is it ensured that the receiver is always dropped?
If the receiver is not dropped, then we would keep sending updates even if the target state was already reched. 
I feel breaking the loop once a reached target was reached was more logical.

Reply:
> Is it ensured that the receiver is always dropped?

At the latest, the receivers will be dropped once we exit from the swap function (because the subscription is not returned). Generally though, as long as the receiver is alive, some code is holding on to the subscription which means some code part is apparently still interested in status changes of this transaction. That in turn means we should keep sending updates.

> I feel breaking the loop once a reached target was reached was more logical - I hope I did not misunderstand the concept.

The idea of this refactoring is/was that instead of calling this function from several places, we only call it once and then use the subscription to check how far we have progressed. That design implies that this loop can no longer know what the target state is.

Instead of the diff, maybe a look at the final result can clear things up. With all the deleted code in the wallet, it is a bit hard to spot how things work now.

---

Your suggestion made me realize that we should "unsubscribe" once the receivers are dropped, otherwise we keep polling updates about a script forever.